File: sampleapplication/src/main/java/com/byoutline/sampleapplication/SampleApp.java
package com.byoutline.sampleapplication;

import android.support.annotation.VisibleForTesting;

import com.byoutline.sampleapplication.di.AppComponent;
import com.byoutline.sampleapplication.di.AppModule;
import com.byoutline.sampleapplication.di.DaggerAppComponent;
import com.byoutline.sampleapplication.di.DaggerGlobalComponent;
import com.byoutline.sampleapplication.di.GlobalComponent;
import com.byoutline.sampleapplication.di.GlobalModule;
import com.byoutline.secretsauce.BaseApp;
import com.squareup.otto.Bus;

/**
 * Application entry point that configures Dagger and SecretSauce.
 */
public class SampleApp extends BaseApp {
    /**
     * Although unusually public static variable is discouraged we keep it,
     * since calls to component look much better as
     * <code>SampleApp.component.inject(this)</code>
     * <p>than alternative:
     * <code>SampleApp.getInstance(context).getComponent().inject(this)</code></p>
     */
    public static GlobalComponent component;


    @Override
    public void onCreate() {
        super.onCreate();
        resetComponents();
    }

    @Override
    protected boolean isDebug() {
        return BuildConfig.DEBUG;
    }

    @VisibleForTesting
    public void resetComponents() {
        AppComponent appComponent = createAppComponent();
        GlobalComponent mainComponent = createGlobalComponent(appComponent.getBus());
        setComponents(mainComponent, appComponent);
    }

    private GlobalComponent createGlobalComponent(Bus bus) {
        return DaggerGlobalComponent.builder()
                .globalModule(new GlobalModule(this, bus))
                .build();
    }

    private AppComponent createAppComponent() {
        return DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();
    }

    @VisibleForTesting
    public synchronized void setComponents(GlobalComponent mainComponent, AppComponent appComponent) {
        component = mainComponent;
        init(appComponent);
    }
}

File: sampleapplication/src/main/java/com/byoutline/sampleapplication/SampleApp.kt
package com.byoutline.sampleapplication

import android.app.Application
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.ViewModelProvider
import com.byoutline.sampleapplication.di.AppComponent
import com.byoutline.sampleapplication.di.DaggerAppComponent
import com.byoutline.sampleapplication.networkchangereceiver.NetworkActivityModule
import com.byoutline.secretsauce.SecretSauceSettings
import com.byoutline.secretsauce.di.AppInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import javax.inject.Inject

/**
 * Application entry point that configures Dagger and SecretSauce.
 */
open class SampleApp : Application(), HasAndroidInjector {

    @Inject lateinit var _androidInjector: DispatchingAndroidInjector<Any>
    override fun androidInjector() = _androidInjector
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    override fun onCreate() {
        super.onCreate()
        initDagger()
    }

    @VisibleForTesting open fun initDagger() {
        setComponents(createGlobalComponent())
    }

    private fun createGlobalComponent(): AppComponent {
        return DaggerAppComponent.builder()
            .networkActivityModule(NetworkActivityModule(this))
            .build()
    }

    @VisibleForTesting
    @Synchronized
    open fun setComponents(mainComponent: AppComponent) {
        mainComponent.inject(this)
        AppInjector.init(this)
        SecretSauceSettings.set(
            debug = BuildConfig.DEBUG,
            containerViewId = R.id.container,
            bindingViewModelId = BR.viewModel,
            viewModelFactoryProvider = { viewModelFactory }
        )
    }
}
